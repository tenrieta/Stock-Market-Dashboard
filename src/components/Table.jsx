import React, { useState } from "react";
import FormControl from "react-bootstrap/FormControl";
import { useDispatch, useSelector } from "react-redux";
import {
  // makeStyles,
  TableBody,
  TableRow,
  TableCell,
  Toolbar,
  InputAdornment,
} from "@material-ui/core";
import { Link } from "react-router-dom";
import Button from "react-bootstrap/Button";
import { Search } from "@material-ui/icons";
import selectStock from "../actions/chosenStockActions";
import useTable from "../hooks/useTable";
import symbols from "../../data/stocks";
import ModalForm from "./ModalForm";
import { getStockDataFromStorage, getYesterday } from "./Chart";

// Redux action creators
import addStockData from "../actions/stockDataActions";
/*
const useStyles = makeStyles((theme) => ({
  pageContent: {
    margin: theme.spacing(5),
    padding: theme.spacing(3),
  },
  searchInput: {
    width: "75%",
  },
}));
*/

const headCells = [
  { id: "symbol", label: "Symbol" },
  { id: "name", label: "Name" },
  { id: "price", label: "Price" },
  { id: "change", label: "Change" },
  { id: "volume", label: "Volume" },
  { id: "actions", label: "Action" },
];

const Table = function Table() {
  const selectedStock = useSelector((state) => state.chosenStock);
  const currentUser = useSelector((state) => state.currentUser);
  const stockDataFromRedux = useSelector((state) => state.stockData);
  // const [stocks] = useState([...symbols]);
  // const classes = useStyles();
  // const [records] = useState([...symbols]);
  // const [symbol, setSymbol] = useState("");
  const [filterFn, setFilterFn] = useState({
    fn: (stocks) => stocks,
  });
  const [modalStatus, setModalStatus] = React.useState({
    modalOpened: false,
    activeStock: null,
  });

  const newSymbols = [...symbols];

  Object(symbols).map((stock) => {
    const stockDataFromStorage1 = getStockDataFromStorage();
    Object.assign(stock, {
      price:
        Math.round(
          Number(
            stockDataFromStorage1?.[Object(stock.symbol)]?.seriesCandle[0]
              ?.data[0]?.y[3]
          ) * 100
        ) / 100,
    });
    Object.assign(stock, {
      change:
        Math.round(
          Math.abs(
            100 -
              (Number(
                stockDataFromStorage1?.[Object(stock.symbol)]?.seriesCandle[0]
                  ?.data[1]?.y[3]
              ) /
                Number(
                  stockDataFromStorage1?.[Object(stock.symbol)]?.seriesCandle[0]
                    ?.data[0]?.y[3]
                )) *
                100
          ) * 100
        ) / 100,
    });
    Object.assign(stock, {
      volume: Math.round(
        Number(
          stockDataFromStorage1?.[Object(stock.symbol)]?.seriesBar[0]?.data[0]
            ?.y
        ) / 1000000
      ),
    });
    return true;
  });

  const { TblContainer, TblHead, TblPagination, recordsAfterPagingAndSorting } =
    useTable(newSymbols, headCells, filterFn);

  const handleSearch = (e) => {
    setFilterFn({
      fn: (stocks) => {
        if (e.target.value === "") return stocks;
        return stocks.filter(
          (x) =>
            x.name.toLowerCase().includes(e.target.value) ||
            x.symbol.toLowerCase().includes(e.target.value)
        );
      },
    });
  };

  const dispatch = useDispatch();

  const handleStockAction = (activeStock, action) => {
    setModalStatus({
      modalOpened: true,
      activeStock,
      action,
    });
  };

  React.useEffect(() => {
    const stockDataFromStorage = getStockDataFromStorage();
    const yesterdayDate = getYesterday();
    symbols.map((stock) => {
      if (
        stockDataFromStorage?.[stock.symbol]?.seriesCandle?.[0]?.data?.[0]
          ?.x === yesterdayDate ||
        stockDataFromStorage?.[stock.symbol]?.seriesCandle?.[0]?.data?.[1]
          ?.x === yesterdayDate
      ) {
        dispatch(
          addStockData(stock.symbol, stockDataFromStorage[stock.symbol])
        );
      }
      return true;
    });
  }, [dispatch]);

  return (
    <>
      {/* Stocks Table */}
      <Toolbar>
        <FormControl
          type="search"
          placeholder="Search"
          aria-label="Search"
          inputprops={{
            startAdornment: (
              <InputAdornment position="start">
                <Search />
              </InputAdornment>
            ),
          }}
          onChange={handleSearch}
        />
      </Toolbar>
      <TblContainer>
        <TblHead />
        <TableBody>
          {recordsAfterPagingAndSorting().map((stock) => (
            <TableRow
              key={stock.id}
              selected={stock.symbol === selectedStock}
              onClick={() => dispatch(selectStock(stock.symbol))}
              id={stock.symbol.concat("-row")}
            >
              <TableCell>{stock.symbol}</TableCell>
              <TableCell>{stock.name}</TableCell>
              {stockDataFromRedux?.[stock.symbol]?.seriesCandle[0]?.data[0] ? (
                <>
                  <TableCell>
                    $
                    {Math.round(
                      Number(
                        stockDataFromRedux?.[stock.symbol]?.seriesCandle[0]
                          ?.data[0]?.y[3]
                      ) * 100
                    ) / 100}
                  </TableCell>
                  <TableCell>
                    {Math.round(
                      Math.abs(
                        100 -
                          (Number(
                            stockDataFromRedux?.[stock.symbol]?.seriesCandle[0]
                              ?.data[1]?.y[3]
                          ) /
                            Number(
                              stockDataFromRedux?.[stock.symbol]
                                ?.seriesCandle[0]?.data[0]?.y[3]
                            )) *
                            100
                      ) * 100
                    ) / 100}
                    %
                  </TableCell>
                  <TableCell>
                    {Math.round(
                      Number(
                        stockDataFromRedux?.[stock.symbol]?.seriesBar[0]
                          ?.data[0]?.y
                      ) / 1000000
                    )}
                    m
                  </TableCell>
                  <TableCell>
                    {currentUser.uid ? (
                      <Button
                        variant={
                          currentUser.firestore?.stockUnits?.[stock.symbol]
                            ? "primary"
                            : "outline-primary"
                        }
                        size="sm"
                        onClick={() => handleStockAction(stock.symbol)}
                      >
                        Buy / Sell
                      </Button>
                    ) : (
                      <Button
                        variant="outline-primary"
                        size="sm"
                        as={Link}
                        to="/log-in"
                      >
                        Buy / Sell
                      </Button>
                    )}
                  </TableCell>
                </>
              ) : (
                <>
                  <TableCell />
                  <TableCell />
                  <TableCell />
                  <TableCell />
                </>
              )}
            </TableRow>
          ))}
        </TableBody>
      </TblContainer>
      <TblPagination />
      <ModalForm modalStatus={modalStatus} setModalStatus={setModalStatus} />
    </>
  );
};

export default Table;
