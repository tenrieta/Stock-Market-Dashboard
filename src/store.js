/* eslint-disable no-underscore-dangle */
import { applyMiddleware, compose, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers";

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk), // Thunk is necessary for async action creators
    window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : (f) => f // This is to make the Redux for Chrome extension work in an incognito window with disabled Redux extension
  )
);

export default store;
