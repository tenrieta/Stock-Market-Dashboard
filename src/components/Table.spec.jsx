/* eslint-disable no-unused-vars */
import React from "react";
import renderer from "react-test-renderer";
import Table from "./Table";
import { Provider } from "react-redux";
import store from "../store";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { FormControl, Toolbar } from "@material-ui/core";
import { shallow, mount } from "enzyme";
import userEvent from "@testing-library/user-event";

test("Always true test", () => {
  expect(true).toBe(true);
});

describe("Table", () => {
  test("Table snapshot renders", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Table />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should have one Table", () => {
    const container = shallow(
      <Provider store={store}>
        <Table />
      </Provider>
    );
    expect(container.find("Table").length).toEqual(1);
  });
});

describe("Input value for Search", () => {
  it("updates on change", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Table />
      </Provider>
    );

    const handleSearch = jest.fn((e) => {
      e.target.value = "Facebook";
    });

    const { queryByPlaceholderText } = render(
      <FormControl type="search" onChange={handleSearch} />
    );
    const searchInput = queryByPlaceholderText("Search");

    //userEvent.type(searchInput, "Facebook");
    expect(handleSearch).not.toHaveBeenCalled();
    //expect(searchInput.value).toBe("Facebook");
  });
});
