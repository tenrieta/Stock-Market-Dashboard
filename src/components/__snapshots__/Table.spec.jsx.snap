// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Table snapshot renders 1`] = `
Array [
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <input
      aria-label="Search"
      className="form-control"
      inputprops={
        Object {
          "startAdornment": <WithStyles(ForwardRef(InputAdornment))
            position="start"
          >
            <Memo />
          </WithStyles(ForwardRef(InputAdornment))>,
        }
      }
      onChange={[Function]}
      placeholder="Search"
      type="search"
    />
  </div>,
  <table
    className="MuiTable-root makeStyles-table-1"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Symbol
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Name
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Price
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Change
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Volume
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root Mui-selected"
        id="SNAP-row"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          SNAP
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Snap Inc.
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
      </tr>
      <tr
        className="MuiTableRow-root"
        id="TWTR-row"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          TWTR
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Twitter, Inc.
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
      </tr>
      <tr
        className="MuiTableRow-root"
        id="AAPL-row"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          AAPL
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Apple Inc.
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
      </tr>
      <tr
        className="MuiTableRow-root"
        id="MSFT-row"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          MSFT
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Microsoft Corporation
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
      </tr>
      <tr
        className="MuiTableRow-root"
        id="CSCO-row"
        onClick={[Function]}
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          CSCO
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          Cisco Systems, Inc.
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        />
      </tr>
    </tbody>
  </table>,
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          5
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={5}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-5 of 10
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
